cpp/src/plasma/tf_plasma_op.cc

```diff
@@ -0,0 +1,383 @@
+    }
+
+    float* data = reinterpret_cast<float*>(data_buffer->mutable_data());
+
+    auto wrapped_callback = [this, context, done, data_buffer, object_id]() {
+      {
+        mutex_lock lock(mu_);
+        ARROW_CHECK_OK(client_.Seal(object_id));
+      }
+      context->SetStatus(tensorflow::Status::OK());
+      done();
+    };
+
+    if (std::is_same<Device, CPUDevice>::value) {
+      for (int i = 0; i < num_tensors; ++i) {
+        const auto& input_tensor = context->input(i);
+        std::memcpy(static_cast<void*>(data + offsets[i] / sizeof(float)),
+                    input_tensor.flat<float>().data(),
+                    static_cast<uint64>(offsets[i + 1] - offsets[i]));
+      }
+      wrapped_callback();
+    } else {
+#ifdef GOOGLE_CUDA
+      auto orig_stream = context->op_device_context()->stream();
+      OP_REQUIRES_ASYNC(context, orig_stream != nullptr,
+                        errors::Internal("No GPU stream available."), done);
+      auto stream_executor = orig_stream->parent();
+
+      // NOTE(zongheng): this is critical of getting good performance out of D2H
+      // async memcpy.  Under the hood it performs cuMemHostRegister(), see:
+      // http://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gf0a9fe11544326dabd743b7aa6b54223
+      CHECK(stream_executor->HostMemoryRegister(static_cast<void*>(data),
+                                                static_cast<uint64>(total_bytes)));
+
+      {
+        mutex_lock l(d2h_stream_mu);
+        if (d2h_stream == nullptr) {
+          d2h_stream = new Stream(stream_executor);
+          CHECK(d2h_stream->Init().ok());
+        }
+      }
+
+      // Needed to make sure the input buffers have been computed.
+      // NOTE(ekl): this is unnecessary when the op is behind a NCCL allreduce already
+      CHECK(d2h_stream->ThenWaitFor(orig_stream).ok());
+
+      for (int i = 0; i < num_tensors; ++i) {
+        const auto& input_tensor = context->input(i);
+        float* input_buffer = const_cast<float*>(input_tensor.flat<float>().data());
+        perftools::gputools::DeviceMemoryBase wrapped_src(
+            static_cast<void*>(input_buffer));
+        const bool success =
+            d2h_stream
+                ->ThenMemcpy(static_cast<void*>(data + offsets[i] / sizeof(float)),
+                             wrapped_src,
+                             static_cast<uint64>(offsets[i + 1] - offsets[i]))
+                .ok();
+        OP_REQUIRES_ASYNC(context, success,
+                          errors::Internal("D2H memcpy failed to be enqueued."), done);
+      }
+      // TODO(zongheng): does std::move() give better performance?
+      context->device()->tensorflow_gpu_device_info()->event_mgr->ThenExecute(
+          d2h_stream, std::move(wrapped_callback));
+#endif
+    }
+  }
+
+ private:
+  std::string plasma_store_socket_name_;
+  std::string plasma_manager_socket_name_;
+
+  mutex mu_;
```

Comment:
Please document all of the private fields. In particular, `mu_` guards more than just `client_`, right?